engines/dfc_provider/app/services/dfc_io.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+# Our interface to the DFC Connector library.
+module DfcIo
+  CONTEXT = JSON.parse <<~JSON
+    {
+      "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
+      "skos" : "http://www.w3.org/2004/02/skos/core#",
+      "dfc": "https://github.com/datafoodconsortium/ontology/releases/latest/download/DFC_FullModel.owl#",
+      "dc": "http://purl.org/dc/elements/1.1/#",
+      "dfc-b": "https://github.com/datafoodconsortium/ontology/releases/latest/download/DFC_BusinessOntology.owl#",
+      "dfc-p": "https://github.com/datafoodconsortium/ontology/releases/latest/download/DFC_ProductGlossary.owl#",
+      "dfc-t": "https://github.com/datafoodconsortium/ontology/releases/latest/download/DFC_TechnicalOntology.owl#",
+      "dfc-m": "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/measures.rdf#",
+      "dfc-pt": "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf#",
+      "dfc-f": "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/facets.rdf#",
+      "ontosec": "http://www.semanticweb.org/ontologies/2008/11/OntologySecurity.owl#",
+      "dfc-p:hasUnit":{ "@type":"@id" },
+      "dfc-b:hasUnit":{ "@type":"@id" },
+      "dfc-b:hasQuantity":{ "@type":"@id" },
+      "dfc-p:hasType":{ "@type":"@id" },
+      "dfc-b:hasType":{ "@type":"@id" },
+      "dfc-b:references":{ "@type":"@id" },
+      "dfc-b:referencedBy":{ "@type":"@id" },
+      "dfc-b:offeres":{ "@type":"@id" },
+      "dfc-b:supplies":{ "@type":"@id" },
+      "dfc-b:defines":{ "@type":"@id" },
+      "dfc-b:affiliates":{ "@type":"@id" },
+      "dfc-b:hasCertification":{ "@type":"@id" },
+      "dfc-b:manages":{ "@type":"@id" },
+      "dfc-b:offeredThrough":{ "@type":"@id" },
+      "dfc-b:hasBrand":{ "@type":"@id" },
+      "dfc-b:hasGeographicalOrigin":{ "@type":"@id" },
+      "dfc-b:hasClaim":{ "@type":"@id" },
+      "dfc-b:hasAllergenDimension":{ "@type":"@id" },
+      "dfc-b:hasNutrientDimension":{ "@type":"@id" },
+      "dfc-b:hasPhysicalDimension":{ "@type":"@id" },
+      "dfc:owner":{ "@type":"@id" },
+      "dfc-t:hostedBy":{ "@type":"@id" },
+      "dfc-t:hasPivot":{ "@type":"@id" },
+      "dfc-t:represent":{ "@type":"@id" }
+    }
+  JSON
```

Comment:
> If DFC update their url, we'll need to update this ?

Depends on the URL. Last time, they updated the URL where this document is published. By including the document in the source code, we don't care about that URL any more.

But they also changed URLs contained in this context. They now point to Github. If they change it again then yes, we need to update the context as long as the update was intentional and permanent and other apps are updating as well.

At the same time we also need to update the DFC Connector library because it has all these URLs hardcoded. So, if the DFC updates URLs and we haven't done it yet then at least our code still works with all other apps which haven't updated their code. Then we can update the Connector and the context at once and maybe add other backwards compatibility to communicate with old and new apps.

> If DFC update contxt, we'll need to update this ?

Yes, if they add more to their data models, for example, they may add or rename something in this context. Then we need to update but we also have to update the DFC Connector library again and maybe some other code to deal with these changes.

So basically, yes, we have to update the context but at least our app keeps working with the old version while we haven't updated the context.

> is there an easy way to do this ? or is it a manual process ?

It's currently downloading the file and copying it into the code. I could also imagine to just download the file and load it from within the code. We have a script to download product types and units already. We can add it to that file.

But if the URLs change then we have to update that script as well so that it knows from where to download the new files. :see_no_evil: 

We need to wait for some stability to add more automation. The DFC is currently developed like a research prototype. Updates have to be reviewed carefully and changes be done manually.

Reply:
Thanks for the clarification !